neigh: Clean up stale/untracked non-GC'ed neighbors

borkmann opened the pull request:
See commit msg. We might need custom backport for <=1.10.

Fixes: #17905

borkmann committed:
neigh: Replace mentions of ARP with neighbor entries in log messages

Since it's not only about ARP anymore. Also remove 'neigh' abbreviation
while at it.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
neigh: Add clean up of stale/untracked neighbors after resync

The NodeNeighDiscoveryEnabled() is called after we completed a resync of all
nodes with kubeapi-server. (See initRestore() -> SyncWithK8sFinished() and the
agent's wait in runDaemon() for restoreComplete channel to finish in non-dry
mode.)

Given that, neighLastPingByNextHop map is populated at that time, so when doing
migration of neighbor entries in NodeCleanNeighbors() we can remove unrelevant
ones at the same time to not let garbage neighbor entries pile up in the neighbor
hash table and/or avoid that the kernel needs to do periodic work for them in
case of NTF_EXT_MANAGED ones.

neighLastPingByNextHop holds both the v4 and v6 entries as a string for the key,
so it's enough to just check it there.

Fixes: #17905
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

aditighag reviewed (approved):
Thanks! 🚀 
For older cilium versions, I suppose we would need to backport changes that populate the `neighNextHopByNode{4,6}` map along with invoking `NodeCleanNeighbors` from `syncWithK8sfinished`?

borkmann committed:
neigh: Add runtime test for cleaning stale IPv6 nextHop addresses

Add a test case which simulates agent resync where the agent cleans up stale
neighbor entries that we no longer track due to nodes not being in cluster.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
neigh: Refactor NodeCleanNeighbors into NodeCleanNeighborsLink

Add a variant which does not depend on the config dir but where we can pass
a link directly to it. This is needed for the runtime test given subsequent
calls to NodeCleanNeighbors() won't have any effect as the link's config file
gets removed upon first successful cleanup. Needed for upcoming runtime tests.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
neigh: Add runtime test for cleaning stale IPv4 nextHop addresses

Similar as we do for IPv6, add a test case which simulates agent resync where
the agent cleans up stale neighbor entries that we no longer track due to nodes
not being in cluster. Make sure that we also have IPv4 entries covered.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
